rosette/src/regexp.cc

```diff
@@ -161,7 +161,7 @@ static const char *rcsid =
  * but allows patterns to get big without disasters.
  */
 #define OP(p) (*(p))
-#define NEXT(p) (((*((p) + 1) & 0377) << 8) + *((p) + 2) & 0377)
+#define NEXT(p) ((((*((p) + 1) & 0377) << 8) + *((p) + 2)) & 0377)
```

Comment:
I wish this had a comment explaining what it did.

Reply:
I think I got the parens I added here wrong. I will fix that.

What it appears to do is assemble a 16 bit value from two 8 bit values. The order is interesting as it seems to be doing it in big-endian order. Also interesting is that it does this using the two bytes that lie 1 byte past the point p. I know p is a byte pointer because the only user of this macro invokes it with a char*. 

so:

skip p[0]            (regexp opcode)
MSB = p[1]         (next pointer MSB)
LSB = p[2]          (next pointer LSB)

WORD16 = MSB<<8 + LSB

Comment in the code:
     A node is one char of opcode followed by two chars of "next" pointer.
     "Next" pointers are stored as two 8-bit pieces, high order first.  The
     value is a positive offset from the opcode of the node containing it.
     An operand, if any, simply follows the node.  (Note that much of the
     code generation knows about this implicit relationship.)

     Using two bytes for the "next" pointer is vast overkill for most things,
     but allows patterns to get big without disasters.


This is only used in one place, near line 1054 in regnext() in regexp.cc.

Side note: regexp.cc is a 1993 implementation of a regular expression parser. Would it make sense to abandon it in lieu of a standard implementation? The only use I can see of this regular expression parser is as a primitive "regexpCompare". This in turn is called two places in apropos.rbl. It is not mentioned in the Rosette manual.

Reply:
Actual pointers don't fit in 2 bytes. On my machine a pointer is 8 bytes. I'm not entirely surprised that this code appears to be written for a 16-bit address space, but I honestly don't know how to handle fixing it. The comment noting that "much of the code generation know about this" makes me sad.

Reply:
**Final:** We will remove this implementation of regexp and replace it with std::regex from the c++ standard library. This will **not** be done in this PR, but will be done in a follow-on PR.